### DISCUSSION TITLE
깊은복사와 얕은복사

### DISCUSSION THREAD

**Discussion by WooGie911** [wanted-9th-3team/frontend-study#28] [프론트 기술 질문 공간] [answered]
깊은복사와 얕은복사에 관련된 경험 혹은 개념을 설명해주세요.

**Comment by skyhanull** [comment]
얕은 복사(Shallow Copy)란 객체를 복사할 때 원본 값과 복사된 값이 같은 참조(=메모리 주소)를 가리키는 것이다.
깊은 복사(Deep Copy)는 복사된 객체가 다른 주소를 참조하며 내부의 값만 복사된다.
깊은 복사는 잘 모르겠지만 얕은 복사는 주로 map,filter,spread연산자 사용시 얕은 복사가 이루어 지는 것으로 알고 있습니다.

**Comment by WooGie911** [comment, marked as answer]
요약 :
- 한마디로 요약하자면 얕은 복사는 객체의 참조값(주소 값)을 복사하고, 깊은 복사는 객체의 실제 값을 복사합니다. 


깊은 복사 (Deep Copy)와 얕은 복사 (Shallow Copy) :
- 깊은 복사(deep copy)와 얕은 복사(shallow copy)는 객체를 복사할 때 차이가 나는 개념입니다. 

- 깊은 복사란, 
객체가 참조하는 자신의 속성들까지 모두 새로운 객체로 복사하는 것을 의미합니다. 
이렇게 깊은 복사를 할 경우, 원래 객체와 복사한 객체는 완전히 독립적인 객체가 되기 때문에
원래 객체의 값이 변경되어도 복사한 객체의 값은 영향을 받지 않습니다. 

- 얕은 복사란, 
객체가 참조하는 속성들은 새로운 객체로 복사하지 않고, 참조만 복사하는 것을 의미합니다. 
이렇게 얕은 복사를 할 경우, 원래 객체와 복사한 객체는 참조하고 있는 속성을 공유하게 되기 때문에 
원래 객체의 값이 변경되면 복사한 객체의 값도 같이 변경됩니다. 


JavaScript에서 깊은 복사를 구현하는 방법 :
1. 재귀함수를 이용한 깊은 복사. 
(단점 : 복잡함)

2. Lodash의 cloneDeep 함수. 
(단점 : 라이브러리 다운받아야함)

3. 객체를 JSON.stringify()함수로 문자열로 만든후 JSON.parse()함수로 다시 객체로 형변환. 
(단점 :  느리고 객체가 함수인 경우 undefind로 처리됨)


경험 : 
- setState를 사용하여 state값을 바꿔 주어도 리렌더링이 일어나지 않는 문제를 직면한 경험이 있습니다.
- 원인은 참조된 주소가 바뀌지 않아 state값이 바뀌는 것을 React가 인지하지 못하여 리렌더링이 일어나지 않는 것이었습니다.
- 이 문제를 해결하기 위해 Spread 연산자를 사용하여 깊은복사를 활용해 주소값을 바꿔주어 해결했습니다.

**Comment by michoball** [comment]
깊은 복사와 얕은 복사는 원시 타입이 아닌 참조 타입을 복사 할 때 나타나는 특징입니다. 

얕은 복사는 참조 타입 값의 주소만을 복사하는 것을 말합니다. 때문에 얕은 복사로 복사한 값을 변경하면 원본 값 또한 변경이 이뤄집니다. 

깊은 복사는 값 자체를 복사해 새로운 변수에 할당해줘서 모습은 같지만 아예 새로운 값이 되는 것을 말합니다. 
원시타입은 다른 변수에 선언시 똑같은 값을 가지는 새로운 변수가 생성되는 것이기에 모두 깊은 복사가 이뤄집니다. 
반면 참조 타입의 경우 다양한 방법으로 깊은 복사를 구현할 수 있습니다. 

저는 대체로 spread 문법으로 많이 구현을 하는데 이는 참조 값의 depths가 2차 이상으로 깊어지면 얕은 복사가 된다는 단점이 있습니다.  (Object.assign()도 마찬가지 )

완전히 깊은 복사를 하려면  다른 라이브러리나 재귀함수 또는 JSON.parse 와 JSON.stringify 사용이 있습니다. 

예전에 코딩 테스트를 본 적이 있는데 그때  배열은 얕은 복사가 된다는 점을 간과하고 문제를 풀다가 쉬운 문제 였는데도 불구하고 한참 걸려서 풀었던 적이 있습니다. ㅠㅠ

**Comment by bona373737** [comment]
데이터 타입에 따라 메모리에 저장되는 방식에 차이가 있는데,  
number, string, null, undefined, bool, symbol 같은 원시타입 데이터의 경우 입력 받은 값을 직접 스택 메모리 에 저장을 하고
array, object, function같은 참조타입 데이터의 경우 실제 데이터는 힙 이라는 메모리 공간에 저장해 두고 스택메모리에는 값이 저장된 곳의 주소를 저장합니다. 참조 타입 데이터(객체, 배열)의 변경에 따라 달라지는 메모리 영역의 사용을 유연하게 하기 위해 이런 방식으로 저장을 합니다. 

값을 복사할때 원시타입의 데이터는 값을 복사할 경우 새로운 메모리에 값 그 자체를 복사해서 사용하게 되지만 참조타입의 데이터의 경우 스프레드연산자 또는 lodash라이브러리를 활용한 깊은 복사를 하지 않는 경우 새로운 메모리에 동일한 주소값만이 복사된다.

**Comment by javascriptbiggosoo** [comment]
reference value가 저장되는 방식: 변수가 object로 선언되는 순간 주소를 받는데, 그 밸류는 heap 안의 또다른 메모리 주소를 가르키는 밸류.
그래서 다른 변수에 할당한 후 그 사본의 variables를 바꾸면 원본까지 바뀌는데 이를 얕은 복사라고 칭함.
primitive는 사본의 value를 수정하면 주소도 (당연히)바뀌면서 원본이 유지되는데 reference 밸류(힙에 있는 것들)는 바뀔 수 있음.
사본을 수정하면 heap의 실제 value가 바뀌는데 call stack이 가르키는 주소는 그대로.
(아예 call stack의 주소 자체를 바꿔버리는 초기화와는 다르다.)

**Comment by BaekYuri** [comment]
얕은 복사는 객체의 주소 값만 복사하는 것이고 깊은 복사는 객체의 실제 값을 복사합니다.
얕은 복사는 let object = object2 형식으로 복사할 수 있고, 깊은 복사는 ... 연산자를 사용하여 간편하게 사용할 수 있습니다.
